keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+namespace has connectivity, even when the truth table explicitly forbids it", which might, for example, point to a bug in a CNI provider
+related to flagrantly allowing internamespace traffic.  Since it is obvious how such a matrix might be defined in Go, we don't provide a
+code snippet or API example.
+ 
+3. (about 80% done modulo ginkgo implementation details) Rewrite each individual test, reviewing semantics, to be precisely worded (and possibly verbose), and to simply define a specific policy and
+set of 'whitelisted' communication associated with this policy. The whitelisting would be defined as a map of namespace->pods, since all other
+information in the truth table is false.
+ 
+Example:
+
+Initially, to confirm the logical capacity of the builder mechanism for replacing existing tests, a prototype of inplace replacements of NetworkPolicy definitions was done here (prototype) https://gist.github.com/6a62266e0eec2b15e5250bd65daa4faa.  
+Now, this underlying API has (essentially) been implemented in full, and the following repository https://github.com/vmware-tanzu/antrea/tree/master/hack/netpol, demonstrates a working implementation and port of the network policy tests (Most of the tests have been ported).  Each test follows a simple and easy to read pattern such as this:
+
+ ```
+ 	// TODO, consider a SpecBuilder.New(...) style of invocation
+	builder := &utils.NetworkPolicySpecBuilder{}
+	builder = builder.SetName("allow-x-via-pod-and-ns-selector").SetPodSelector(map[string]string{"pod": "a"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod":"b"}, map[string]string{"ns":"y"}, nil, nil)
+
+	reachability := utils.NewReachability(allPods, true)
+	reachability.ExpectAllIngress(Pod("x/a"), false)
+	reachability.Expect(Pod("y/b"), Pod("x/a"), true)
+	reachability.Expect(Pod("x/a"), Pod("x/a"), true)
+  ```
+ This represents a significant reduction in code complexity, with the equivalent tests using the existing `network_policy.go` implementation being 3 to 4 times as long, mostly due to boilerplate around verification and go structures.
+
+*Further improvements to the testing API*
+
+- Make the function calls in network policy builder *even* more DSL-like, for example, 
+```
+Pod(...).InNamespace(...).CanAccess(...)
+```
+- Infer `Egress,Ingress` rules rather than force them to be specified, based on builder inputs.  They're redundant to begin with (i.e. calico doesn't even require them)
+- Add `From` and `To` semantics to the struct API calls in reachability.
+ 
+###  Note on Acceptance and Backwards compatibility
+
+Thus far there are two obvious ways to ensure backwards compatibility. 
+- Each old test will be converted to a truth table, first, as part of this work, so that the parity between old and new tests is clear and obvious 
+- Alternatively, these can be a next generation of policy tests which live in parallel to existing network policy tests for a release cycle, while they are vetted
```

Comment:
I don't think we need to guarantee backward compatibility. We should be able to point to rows of the table and say "this one corresponds to old test case X" and convince ourselves that we didn't lose any coverage.

Reply:
gotcha, fixed, pushing shortly